// hermes-javascript/dist/api/injection/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const ApiSubset_1 = __importDefault(require("../ApiSubset"));
const enums = __importStar(require("../types/enums"));
/**
 * The Injection API subset.
 */
class Injection extends ApiSubset_1.default {
    constructor(protocolHandler, call, options) {
        super(protocolHandler, call, options, 'hermes_protocol_handler_injection_facade');
        this.publishEvents = {
            injection_request: {
                fullEventName: 'hermes_injection_publish_injection_request_json'
            },
            injection_status_request: {
                fullEventName: 'hermes_injection_publish_injection_status_request_json'
            },
            injection_reset_request: {
                fullEventName: 'hermes_injection_publish_injection_reset_request_json'
            }
        };
        this.publishMessagesList = undefined;
        this.subscribeEvents = {
            injection_status: {
                fullEventName: 'hermes_injection_subscribe_injection_status_json'
            },
            injection_complete: {
                fullEventName: 'hermes_injection_subscribe_injection_complete_json'
            },
            injection_reset_complete: {
                fullEventName: 'hermes_injection_subscribe_injection_reset_complete_json'
            }
        };
        this.subscribeMessagesList = undefined;
    }
    destroy() {
        this.call('hermes_drop_injection_facade', this.facade);
    }
}
exports.default = Injection;
/**
 * Injection enumerations.
 */
Injection.enums = {
    injectionKind: enums.injectionKind
};
